sklearn/ensemble/_hist_gradient_boosting/tests/test_binning.py

```diff
@@ -312,3 +319,145 @@ def test_infinite_values():
+    missing_val_bin = n_bins - 1
+    expected_trans = np.array([[5, 1, missing_val_bin, 4, missing_val_bin, 3,
+                                2, missing_val_bin, 0]]).T
+    assert_array_equal(X_trans, expected_trans)
+
+
+@pytest.mark.parametrize(
+    "n_bins, expected_trans, expected_bin_categories",
+    [
+        (3, [2, 2, 2, 2, 2, 1, 2, 2, 0], [0, 7]),
+        # 0 is choosen instead of 10 because it comes before in lexicon
+        # order
+        (4, [2, 3, 3, 3, 3, 1, 3, 3, 0], [0, 7, 10]),
+        (5, [3, 1, 4, 4, 4, 2, 4, 4, 0], [0, 1, 7, 10]),
+        (6, [4, 1, 5, 5, 5, 3, 2, 5, 0], [0, 1, 4, 7, 10]),
+    ])
+def test_categorical_n_bins_less_than_cardinality(
+        n_bins, expected_trans, expected_bin_categories):
+    # test when there is not enough bins: only the top n_bins - 1 distinct
+    # values (based on cardinality) will have their own bin
+    X = np.array([[4] * 2 + [1] * 3 + [10] * 4 +
+                  [0] * 4 + [9] + [7] * 5], dtype=X_DTYPE).T
+    X_test = np.array([[10, 1, -1, 9, np.nan, 7, 4, 100, 0]], dtype=X_DTYPE).T
+
+    bin_mapper = _BinMapper(n_bins=n_bins,
+                            is_categorical=np.array([True])).fit(X)
+    assert bin_mapper.n_bins_non_missing_ == [n_bins - 1]
+    assert_allclose(bin_mapper.bin_categories_[0], expected_bin_categories)
+    X_trans = bin_mapper.transform(X_test)
+
+    # missing, negative, unknown values are mapped to the missing bin
+    expected_trans = np.array([expected_trans]).T
+    assert_array_equal(X_trans, expected_trans)
+
+
+def test_categorical_n_bins_less_than_cardinality_ties():
+    # when there are not enough bins and the cardiality is the same
+    # the elements with a higher lexicon ordering will be choosen as
+    # categories
+    X = np.array([[1] * 4 + [2] * 4 + [3] * 4 +
+                  [4] * 2 + [5] * 2 + [6] * 1], dtype=X_DTYPE).T
+    X_test = np.array([[1, 2, 3, 4, 5, 6, 7]], dtype=X_DTYPE).T
+
+    # With 4 bins used for non missing values. categories 1, 2, 3, 4
+    # will have their own bin, the rest will be placed in the missing bin.
+    bin_mapper = _BinMapper(n_bins=5, is_categorical=np.array([True])).fit(X)
+    assert bin_mapper.n_bins_non_missing_ == [4]
+    assert_allclose(bin_mapper.bin_categories_[0], [1, 2, 3, 4])
+
+    X_trans = bin_mapper.transform(X_test)
+    expected_trans = np.array([[0, 1, 2, 3, 4, 4, 4]]).T
+    assert_array_equal(X_trans, expected_trans)
+
+    # With 3 bins usd for non missing values. categoires 1, 2
+    # will have their own bin, the rest will be placed in the missing bin.
+    bin_mapper = _BinMapper(n_bins=3, is_categorical=np.array([True])).fit(X)
+    assert bin_mapper.n_bins_non_missing_ == [2]
+    assert_allclose(bin_mapper.bin_categories_[0], [1, 2])
+
+    X_trans = bin_mapper.transform(X_test)
+    expected_trans = np.array([[0, 1, 2, 2, 2, 2, 2]]).T
+    assert_array_equal(X_trans, expected_trans)
+
+
+@pytest.mark.parametrize("missing_value", [-1, np.nan])
```

Comment:
maybe we should remove the parametrization and check both cases at the same time.